reset.c

```diff
@@ -0,0 +1,593 @@
+                break;
+            }
+        }
+    }
+    else {        //keep all
+        for (i = 0; i < count; ++i) {
+            if (sam_hdr_find_line_pos(in_samhdr, "PG", i, &line)) {
+                fprintf(stderr, "Failed to get PG data at %d!\n", i + 1);
+                ret = 1;
+                break;
+            }
+            //line has the required PG data
+            if (sam_hdr_add_lines(out_samhdr, line.s, line.l)) {
+                fprintf(stderr, "Failed to add PG data!\n");
+                ret = 1;
+                break;
+            }
+        }
+    }
+
+    if (!ret) {
+        //add PG entry with reset command irrespective of no-PG flag
+        if (-1 == (ret = sam_hdr_add_pg(out_samhdr, "samtools", "CL", argdump, NULL))) {
+            fprintf(stderr, "Failed to set PG entry!\n");
+        }
+    }
+    ks_free(&line);
+    ks_free(&pn);
+
+    return ret;
+}
+
+/// reset - do the reset of data and create output; create output header with required rg/pg data, add bamdata with flags set to unmapped, pair info and orientation reset,
+// reerse and complement alignment if required
+/** @param infile - input samfile pointer
+ *  @param outfile - output sam file pointer
+ *  @param config - pointer to internal configuration data
+ *  @param args - string containing dump of command line invocation
+returns 1 on failure 0 on success
+*/
+int reset(samFile *infile, samFile *outfile, conf_data *config, char *args)
+{
+    sam_hdr_t *in_samhdr = NULL, *out_samhdr = NULL;
+
+    if (!infile || !outfile) {
+        fprintf(stderr, "Invalid parameters in reset!\n");
+        return EXIT_FAILURE;
+    }
+
+    //read input header
+    in_samhdr = sam_hdr_read(infile);
+    if (!in_samhdr)
+    {
+        fprintf(stderr, "Failed to read header from file!\n");
+        return EXIT_FAILURE;
+    }
+    //create output header
+    if (!(out_samhdr = sam_hdr_init()))
+    {
+        fprintf(stderr, "Failed to create output header!\n");
+        sam_hdr_destroy(in_samhdr);
+        return EXIT_FAILURE;
+    }
+    //TODO SO/GO status not maintained, do they hold as secondary n supplementary are removed?
+    //add version to output header
+    if  (-1 == sam_hdr_add_line(out_samhdr,"HD", "VN", SAM_FORMAT_VERSION, NULL)) {
+        fprintf(stderr, "Failed to set header data!\n");
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        return EXIT_FAILURE;
+    }
+    //add RG / PG lines if configured
+    if ((config->keepRGs && getRGlines(in_samhdr, out_samhdr)) ||
+            getPGlines(in_samhdr, out_samhdr, config, args)) {
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        return EXIT_FAILURE;
+    }
+
+    //write output header
+    if (sam_hdr_write(outfile, out_samhdr)) {
+        print_error_errno("reset", "Output header write failed (%d)!\n", errno);
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        return EXIT_FAILURE;
+    }
```

Comment:
Don't feel frightened to use `goto` for error handling.  We're not proud! :-)

While it's questionable in other cases, it can greatly reduce code duplication for cases like this.  Eg move `ret_r` declaratrion to the top of the function and initialise it to `EXIT_FAILURE`, add an `error:` label just before the final cleanup section in this function, and then all the error recovery can be a self-documenting `goto error;` statement.

Obviously this needs to make sure that all memory being freed is initialised to NULL at the start and if appropriate checked for before freeing.  See the plethora of gotos in https://github.com/samtools/samtools/blob/develop/bam_sort.c#L2719-L3101 as an example.

That said - there's nothing at all wrong with what you've written.  Just note we won't complain at use of goto in C when appropriate. :)  (I know it has a bad rap generally!)

Reply:
Updated.